Implement CSP hash reporting for scripts

yoavweiss opened the pull request:
#### 70d6fcb9fc882b766302bdb5116a50f05414383b
<pre>
Implement CSP hash reporting for scripts
<a href="https://bugs.webkit.org/show_bug.cgi?id=285292">https://bugs.webkit.org/show_bug.cgi?id=285292</a>

Reviewed by Darin Adler.

CSP was recently added new `report-sha256`, `report-sha384` and `report-sha512` keywords - <a href="https://github.com/w3c/webappsec-csp/pull/693/files">https://github.com/w3c/webappsec-csp/pull/693/files</a>

These new keywords trigger a new reporting type &quot;hash-report&quot;.
It reports hashes for (same-origin or CORS enabled) scripts that are loaded in the context of the document (regardless of their &quot;integrity&quot; attribute), and sends reports about them.

Those reports enable developers to:

* Create inventory of the scripts running on their page. (critical for PCI-DSS v4 - context.)
* Have certainty that they can enable SRI or CSP hash-based enforcement without breaking their sites. The current PR only covers external scripts. We may want to extend the feature in the future to cover inline scripts, evals, event handlers and javascript URLs.

This PR implements that feature.

* LayoutTests/TestExpectations: Stop skipping the relevant tests.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/default-src.https.window-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/default-src.https.window.js.sub.headers: Avoid domains.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/multiple-policies.https.sub-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/multiple-policies.https.sub.html.sub.headers: Avoid domains.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/reportonly-default-src.https.window-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/reportonly-default-src.https.window.js.sub.headers: Avoid domains.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/reportonly-script-src-elem.https.window-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/reportonly-script-src-elem.https.window.js.sub.headers: Avoid domains.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/reportonly-script-src-none.https.window-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/reportonly-script-src-none.https.window.js.sub.headers: Avoid domains.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/reportonly-script-src.https.window-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/reportonly-script-src.https.window.js.sub.headers: Avoid domains.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/resources/report-hash-test-runner.sub.js:
(report_hash_test): Avoid domains and add error handling.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/script-src-elem.https.window-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/script-src-elem.https.window.js.sub.headers: Avoid domains.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/script-src-sha512.https.window-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/script-src-sha512.https.window.js.sub.headers: Avoid domains.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/script-src.https.window-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/script-src.https.window.js.sub.headers: Avoid domains.
* LayoutTests/platform/mac-wk1/TestExpectations: Expect failures on wk1 - <a href="https://bugs.webkit.org/show_bug.cgi?id=285098.">https://bugs.webkit.org/show_bug.cgi?id=285098.</a>
* Source/WebCore/Modules/reporting/ViolationReportType.h: Add
  CSPHashReport type.
* Source/WebCore/bindings/js/ScriptModuleLoader.cpp:
(WebCore::ScriptModuleLoader::notifyFinished): rename
matchIntegrityMetadata call.
* Source/WebCore/dom/LoadableClassicScript.cpp:
(WebCore::LoadableNonModuleScriptBase::notifyFinished): rename
matchIntegrityMetadata call.
* Source/WebCore/html/HTMLLinkElement.cpp:
(WebCore::HTMLLinkElement::setCSSStyleSheet): rename
matchIntegrityMetadata call.
* Source/WebCore/loader/DocumentThreadableLoader.cpp:
(WebCore::DocumentThreadableLoader::didFinishLoading): rename
matchIntegrityMetadata call.
* Source/WebCore/loader/PingLoader.cpp:
(WebCore::PingLoader::sendViolationReport): Handle CSPHashReport.
* Source/WebCore/loader/SubresourceIntegrity.cpp:
(WebCore::createReportFormData): Create a report.
(WebCore::addHashPrefix): Add a prefix to the reported value, based on
the algorithm enum value.
(WebCore::findStrongestAlgorithm): Get the strongest algorithm is
a HashAlgorithmSet.
(WebCore::reportHashesIfNeeded): Potentially report hashes for a resource.
(WebCore::matchIntegrityMetadataSlow): implements the
matchIntegrityMetadata logic.
* Source/WebCore/loader/SubresourceIntegrity.h:
(WebCore::matchIntegrityMetadata): Adjust condition and call hash
reporting.
* Source/WebCore/loader/cache/CachedResource.h:
(WebCore::CachedResource::loader const): Make it a const.
(WebCore::CachedResource::setIsHashReportingNeeded): Setter for
isHashReportingNeeded.
(WebCore::CachedResource::isHashReportingNeeded const): Getter for
isHashReportingNeeded.
(WebCore::CachedResource::loader): Deleted.
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::requestResource): Set
isHashReportingNeeded on the resource.
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
(WebCore::ContentSecurityPolicy::hashesToReport): Get the hashes to
report from the different CSP policies.
* Source/WebCore/page/csp/ContentSecurityPolicy.h: Define
  HashAlgorithmSet and HashAlgorithmSetCollection.
* Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.cpp:
(WebCore::ContentSecurityPolicyDirectiveList::hashReportDirectiveForScript const): Get the directives for script hash reporting.
(WebCore::ContentSecurityPolicyDirectiveList::reportHash const): Return
the HashAlgorithmSet for the script reporting directive.
* Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.h: Add
  reportHash and hashReportDirectiveForScript.
* Source/WebCore/page/csp/ContentSecurityPolicySourceList.cpp:
(WebCore::ContentSecurityPolicySourceList::parseSource): Parse the hash
reporting keywords and set the appropriate algorithms in the
HashAlgorithmSet.
* Source/WebCore/page/csp/ContentSecurityPolicySourceList.h: Add
  a HashAlgorithmSet.
(WebCore::ContentSecurityPolicySourceList::reportHash const): Getter for
the HashAlgorithmSet.
* Source/WebCore/page/csp/ContentSecurityPolicySourceListDirective.h:
(WebCore::ContentSecurityPolicySourceListDirective::reportHash const):
Pipe the HashAlgorithmSet from the sourceList.

Canonical link: <a href="https://commits.webkit.org/288506@main">https://commits.webkit.org/288506@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/687543b022ebfd4e63ad4a63e106fd33b1552324

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/83542 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3158 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/37835 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/88613 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/34550 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/85627 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/3246 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11117 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65001 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/22736 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/86588 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/2383 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75908 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45288 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2287 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30122 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/33598 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/73370 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/30852 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/89990 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/10807 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/7798 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73436 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11030 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71728 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/72651 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17981 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16883 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/15598 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/2131 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/10759 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/16231 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/10605 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14081 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/12379 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on LayoutTests/platform/mac-wk1/TestExpectations:
```diff
@@ -860,6 +860,8 @@ webkit.org/b/246719 imported/w3c/web-platform-tests/reporting/generateTestReport
 webkit.org/b/246719 imported/w3c/web-platform-tests/reporting/cross-origin-report-no-credentials.https.sub.html [ Failure ]
 webkit.org/b/246719 imported/w3c/web-platform-tests/reporting/cross-origin-reports-isolated.https.sub.html [ Failure ]
 
+webkit.org/b/285098 imported/w3c/web-platform-tests/content-security-policy/report-hash/ [ Failure ]
```
Can we include a comment explaining why this is here?

yoavweiss replied:
Done

darinadler commented on Source/WebCore/Modules/reporting/ViolationReportType.h:
```diff
@@ -31,6 +31,7 @@ enum class ViolationReportType : uint8_t {
     CORPViolation, // https://fetch.spec.whatwg.org/#queue-a-cross-origin-embedder-policy-corp-violation-report
     ContentSecurityPolicy,
+    CSPHashReport,
     CrossOriginOpenerPolicy,
```
These were sorted case sensitive alphabetical before, not logically. So CSP should come before Content.

yoavweiss replied:
Done

darinadler commented on Source/WebCore/bindings/js/ScriptModuleLoader.cpp:
```diff
@@ -508,7 +508,7 @@ void ScriptModuleLoader::notifyFinished(ModuleScriptLoader& moduleScriptLoader,
 
         if (!integrity.isEmpty()) {
-            if (!matchIntegrityMetadata(cachedScript, integrity)) {
+            if (!matchIntegrityMetadataAndReportHashIfNeeded(cachedScript, integrity)) {
```
Do we need this longer name? Are there callers that don't want the _report hash_ behavior?

yoavweiss replied:
I don't think there are callers that don't want to report the hash, I just thought that it's better to call out the extra functionality in the function name. Happy to revert if you think this is not necessary.

darinadler replied:
Yes, I think the name change is not an improvement. I can see how it might be best to be explicit about this side effect, but the longer name ends up so long I don't think it adds clarity even though the short name is inaccurate. Naming remains an art, not science.

yoavweiss replied:
OK, reverted!

darinadler commented on Source/WebCore/loader/PingLoader.cpp:
```diff
@@ -184,6 +184,7 @@ void PingLoader::sendViolationReport(LocalFrame& frame, const URL& reportURL, Re
         break;
     case ViolationReportType::COEPInheritenceViolation:
+    case ViolationReportType::CSPHashReport:
     case ViolationReportType::CORPViolation:
```
Case sensitive alphabetical sorting would put CSP after CORP. Also noticed that _inheritance_ is misspelled here, with an _e_ instead of an _a_.

yoavweiss replied:
Done. I suspect it'd be better to fix the inheritance typo in another PR, as it seems like it is permeated across file names as well.

darinadler commented on Source/WebCore/loader/SubresourceIntegrity.cpp:
```diff
@@ -161,6 +162,87 @@ static Vector<EncodedResourceCryptographicDigest> strongestMetadataFromSet(Vecto
+    reportObject->setObject("body"_s, WTFMove(body));
+    reportObject->setString("type"_s, type);
+    reportObject->setString("user_agent"_s, userAgent);
+    reportObject->setInteger("age"_s, 0); // We currently do not delay sending the reports.
```
This comment is mysterious and needn't be. Comments need to say why the code does what it does, not just what it does. Why don't we delay? Why does the specification allow for delay? Do we need to change this in the future? Need to touch on this to make it a good comment, but still good to be terse if we can.

yoavweiss replied:
Changed it to indicate reasoning with a pointer to the relevant spec.

darinadler commented on Source/WebCore/loader/SubresourceIntegrity.cpp:
```diff
@@ -161,6 +162,87 @@ static Vector<EncodedResourceCryptographicDigest> strongestMetadataFromSet(Vecto
+    return FormData::create(reportList->toJSONString().utf8());
+}
+
+static String addHashPrefix(ResourceCryptographicDigest::Algorithm algorithm, const String& hash)
```
```suggestion
static String addHashPrefix(ResourceCryptographicDigest::Algorithm algorithm, StringView hash)
```

yoavweiss replied:
done!

darinadler commented on Source/WebCore/loader/SubresourceIntegrity.cpp:
```diff
@@ -161,6 +162,87 @@ static Vector<EncodedResourceCryptographicDigest> strongestMetadataFromSet(Vecto
+    return String();
+}
+
+static std::optional<ResourceCryptographicDigest::Algorithm> getStrongestAlgorithm(HashAlgorithmSet algorithmSet)
```
```suggestion
static std::optional<ResourceCryptographicDigest::Algorithm> findStrongestAlgorithm(HashAlgorithmSet algorithmSet)
```
WebKit coding style does not use _get_ for function names like this, reserving it for functions with "out arguments". Leaving the verb _find_ is OK, though. Or it could just be named `strongestAlgorithm`.

yoavweiss replied:
done!

darinadler commented on Source/WebCore/loader/SubresourceIntegrity.cpp:
```diff
@@ -161,6 +162,87 @@ static Vector<EncodedResourceCryptographicDigest> strongestMetadataFromSet(Vecto
+
+static std::optional<ResourceCryptographicDigest::Algorithm> getStrongestAlgorithm(HashAlgorithmSet algorithmSet)
+{
+    for (int i = 2; i >= 0; --i) {
```
Magic number `2` here is mysterious. Can we write this without the magic number?

yoavweiss replied:
Yup!!

darinadler commented on Source/WebCore/loader/SubresourceIntegrity.cpp:
```diff
@@ -161,6 +162,87 @@ static Vector<EncodedResourceCryptographicDigest> strongestMetadataFromSet(Vecto
+    if (!resource.loader() || !resource.loader()->frame() || !resource.loader()->frame()->document())
+        return;
+
+    RefPtr<Document> document(resource.loader()->frame()->document());
```
```suggestion
    RefPtr document = resource.loader()->frame()->document();
```

yoavweiss replied:
done

darinadler commented on Source/WebCore/loader/SubresourceIntegrity.cpp:
```diff
@@ -161,6 +162,87 @@ static Vector<EncodedResourceCryptographicDigest> strongestMetadataFromSet(Vecto
+    auto csp = document->checkedContentSecurityPolicy();
+    URL documentURL = document->url();
+
+    const HashAlgorithmSetCollection* hashesToReport = csp->hashesToReport();
```
```suggestion
    auto* hashesToReport = csp->hashesToReport();
```

yoavweiss replied:
done

darinadler commented on Source/WebCore/loader/SubresourceIntegrity.cpp:
```diff
@@ -161,6 +162,87 @@ static Vector<EncodedResourceCryptographicDigest> strongestMetadataFromSet(Vecto
+    for (auto pair : *hashesToReport) {
+        HashAlgorithmSet algorithmSet = pair.first;
+        Vector<String>& endpoints = pair.second;
+        std::optional<ResourceCryptographicDigest::Algorithm> hashAlgorithm = getStrongestAlgorithm(algorithmSet);
```
```suggestion
        auto algorithm = findStrongestAlgorithm(algorithmSet);
```

yoavweiss replied:
done

darinadler commented on Source/WebCore/loader/SubresourceIntegrity.cpp:
```diff
@@ -161,6 +162,87 @@ static Vector<EncodedResourceCryptographicDigest> strongestMetadataFromSet(Vecto
+        String hash = ""_s;
+        if (canExposeHashes)
+            hash = addHashPrefix(hashAlgorithm.value(), base64EncodeToString(resource.cryptographicDigest(hashAlgorithm.value()).value));
+        Ref<FormData> report = createReportFormData("csp-hash"_s, documentURL, document->httpUserAgent(), [&](auto& body) {
```
```suggestion
        Ref report = createReportFormData("csp-hash"_s, documentURL, document->httpUserAgent(), [&](auto& body) {
```

yoavweiss replied:
done!

darinadler commented on Source/WebCore/loader/SubresourceIntegrity.cpp:
```diff
@@ -161,6 +162,87 @@ static Vector<EncodedResourceCryptographicDigest> strongestMetadataFromSet(Vecto
+        });
+        document->sendReportToEndpoints(documentURL, Vector<String>(), endpoints, WTFMove(report), ViolationReportType::CSPHashReport);
+
+    }
```
```suggestion
        document->sendReportToEndpoints(documentURL, { }, endpoints, WTFMove(report), ViolationReportType::CSPHashReport);
    }
```

yoavweiss replied:
done

darinadler commented on Source/WebCore/loader/cache/CachedResource.h:
```diff
@@ -24,6 +24,7 @@
 
 #include "CacheValidation.h"
 #include "CachedResourceClient.h"
+#include "ContentSecurityPolicy.h"
```
Would be better to factor this so we don't have to add the include to this header, if we can. I think it should be added to the .cpp file, not the header.

yoavweiss replied:
Removed it entirely, as this was a spurious leftover

darinadler commented on Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
```diff
@@ -1189,4 +1189,15 @@ void ContentSecurityPolicy::setInsecureNavigationRequestsToUpgrade(HashSet<Secur
+        m_hashesToReport = WTF::makeUnique<HashAlgorithmSetCollection>();
+        for (auto& policy : m_policies) {
+            if (auto hash = policy->reportHash())
+                m_hashesToReport->append(std::make_pair(static_cast<uint8_t>(hash), policy->reportToTokens()));
```
Is this `static_cast` needed? If so, why?

yoavweiss replied:
It's not! removed

darinadler commented on Source/WebCore/page/csp/ContentSecurityPolicy.h:
```diff
@@ -290,6 +294,7 @@ class ContentSecurityPolicy final : public CanMakeThreadSafeCheckedPtr<ContentSe
     mutable std::optional<ContentSecurityPolicyResponseHeaders> m_cachedResponseHeaders;
     bool m_isHeaderDelivered { false };
     ContentSecurityPolicyModeForExtension m_contentSecurityPolicyModeForExtension { ContentSecurityPolicyModeForExtension::None };
+    std::unique_ptr<HashAlgorithmSetCollection> m_hashesToReport;
```
Since this never needs to be resized after being created it should be a `FixedVector` (not sure I remembered that name right) and then doesn't need to be stored in a `unique_ptr` since an empty one is just a null pointer I think. The main feature of `Vector` is its ability to be efficiently resized.

yoavweiss replied:
Apologies, but I don't think I understand. In ContentSecurityPolicy::hashesToReport, we are appending pairs to the initially empty vector (so IIUC, it does get resized from its initial empty state). Are you suggesting a different approach?

darinadler replied:
When creating we can use something resizable like a `Vector` with inline capacity then copy into a fixed vector to keep it around. `Vector` pays extra memory so it can be efficiently resized later, a capability we don't need. And it's particularly wasteful to use a Vector stored on the heap.

yoavweiss replied:
Got it! done!

darinadler commented on Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.cpp:
```diff
@@ -776,4 +776,19 @@ bool ContentSecurityPolicyDirectiveList::shouldReportSample(const String& violat
+
+HashAlgorithmSet ContentSecurityPolicyDirectiveList::reportHash() const
+{
+    const ContentSecurityPolicySourceListDirective* directive = hashReportDirectiveForScript();
```
```suggestion
    auto* directive = hashReportDirectiveForScript();
```

yoavweiss replied:
done

darinadler commented on Source/WebCore/page/csp/ContentSecurityPolicySourceList.h:
```diff
@@ -60,6 +60,9 @@ class ContentSecurityPolicySourceList {
 
+    HashAlgorithmSet reportHash() const { return m_reportHash; }
+
+
```
```suggestion
    HashAlgorithmSet reportHash() const { return m_reportHash; }

```

darinadler replied:
I see this was unclear. I was suggesting one blank line rather than two.

yoavweiss replied:
done

yoavweiss reviewed (commented):
Thanks for reviewing!!

darinadler commented on Source/WebCore/loader/SubresourceIntegrity.cpp:
```diff
@@ -161,6 +162,89 @@ static Vector<EncodedResourceCryptographicDigest> strongestMetadataFromSet(Vecto
+    if (!resource.loader() || !resource.loader()->frame() || !resource.loader()->frame()->document())
+        return;
+
+    RefPtr<Document> document = resource.loader()->frame()->document();
```
```suggestion
    RefPtr document = resource.loader()->frame()->document();
```

yoavweiss replied:
done

darinadler commented on Source/WebCore/loader/SubresourceIntegrity.cpp:
```diff
@@ -161,6 +162,89 @@ static Vector<EncodedResourceCryptographicDigest> strongestMetadataFromSet(Vecto
+    bool canExposeHashes = isResponseEligible(resource);
+    for (auto pair : *hashesToReport) {
+        HashAlgorithmSet algorithmSet = pair.first;
+        Vector<String>& endpoints = pair.second;
```
```suggestion
    for (auto& [algorithmSet, endpoints] : *hashesToReport) {
```

yoavweiss replied:
done!

darinadler commented on Source/WebCore/loader/SubresourceIntegrity.cpp:
```diff
@@ -161,6 +162,89 @@ static Vector<EncodedResourceCryptographicDigest> strongestMetadataFromSet(Vecto
+            body.setString("destination"_s, "script"_s);
+        });
+        document->sendReportToEndpoints(documentURL, { }, endpoints, WTFMove(report), ViolationReportType::CSPHashReport);
+
```
Remove this blank line?

yoavweiss replied:
removed

darinadler commented on Source/WebCore/page/csp/ContentSecurityPolicy.h:
```diff
@@ -230,6 +234,7 @@ class ContentSecurityPolicy final : public CanMakeThreadSafeCheckedPtr<ContentSe
     const String& webAssemblyErrorMessage() const { return m_lastPolicyWebAssemblyDisabledErrorMessage; }
 
     ContentSecurityPolicyModeForExtension contentSecurityPolicyModeForExtension() const { return m_contentSecurityPolicyModeForExtension; }
+    const HashAlgorithmSetCollection* hashesToReport();
```
```suggestion
    const HashAlgorithmSetCollection& hashesToReport();
```

yoavweiss replied:
done

darinadler commented on Source/WebCore/page/csp/ContentSecurityPolicy.h:
```diff
@@ -87,6 +88,9 @@ enum class AllowTrustedTypePolicy : uint8_t {
 };
 
+using HashAlgorithmSet = uint8_t;
+using HashAlgorithmSetCollection = FixedVector<std::pair<HashAlgorithmSet, Vector<String>>>;
```
Occurs to me that this could also use `FixedVector<String>`.

yoavweiss replied:
Changed it! This also required a conversion back from FixedVector to Vector before sending the report. I created a custom one, as I don't know if that's common enough to bake in to Vector

darinadler replied:
As a follow up we can fix that by changing `sendReportToEndpoints` to take a `std::span<const String>` instead of a `const Vector<String>&`.

darinadler commented on Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.cpp:
```diff
@@ -776,4 +776,19 @@ bool ContentSecurityPolicyDirectiveList::shouldReportSample(const String& violat
+    if (!operativeDirective || !operativeDirective->reportHash())
+        return nullptr;
+
+    return operativeDirective;
```
```suggestion
    auto* directive = operativeDirectiveScript(m_scriptSrcElem.get(), ContentSecurityPolicyDirectiveNames::scriptSrcElem);
    if (!directive || !directive->reportHash())
        return nullptr;
    return directive;
```

yoavweiss replied:
Done

darinadler commented on Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
```diff
@@ -1189,4 +1189,16 @@ void ContentSecurityPolicy::setInsecureNavigationRequestsToUpgrade(HashSet<Secur
+            if (auto hash = policy->reportHash())
+                hashesToReport.append(std::make_pair(hash, policy->reportToTokens()));
+        }
+        m_hashesToReport = hashesToReport;
```
This should be `m_hashesToReport =  WTFMove(hashesToReport)` for better efficiency, so we move the fixed vectors of strings instead of copying them.

yoavweiss replied:
of course. done!

darinadler commented on Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.cpp:
```diff
@@ -776,4 +776,19 @@ bool ContentSecurityPolicyDirectiveList::shouldReportSample(const String& violat
+    auto* directive = this->operativeDirectiveScript(m_scriptSrcElem.get(), ContentSecurityPolicyDirectiveNames::scriptSrcElem);
+    if (!directive || !directive->reportHash())
+        return nullptr;
+
```
Please omit this blank line.

yoavweiss replied:
done

yoavweiss committed:
Implement CSP hash reporting for scripts
https://bugs.webkit.org/show_bug.cgi?id=285292

Reviewed by Darin Adler.

CSP was recently added new `report-sha256`, `report-sha384` and `report-sha512` keywords - https://github.com/w3c/webappsec-csp/pull/693/files

These new keywords trigger a new reporting type "hash-report".
It reports hashes for (same-origin or CORS enabled) scripts that are loaded in the context of the document (regardless of their "integrity" attribute), and sends reports about them.

Those reports enable developers to:

* Create inventory of the scripts running on their page. (critical for PCI-DSS v4 - context.)
* Have certainty that they can enable SRI or CSP hash-based enforcement without breaking their sites. The current PR only covers external scripts. We may want to extend the feature in the future to cover inline scripts, evals, event handlers and javascript URLs.

This PR implements that feature.

* LayoutTests/TestExpectations: Stop skipping the relevant tests.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/default-src.https.window-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/default-src.https.window.js.sub.headers: Avoid domains.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/multiple-policies.https.sub-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/multiple-policies.https.sub.html.sub.headers: Avoid domains.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/reportonly-default-src.https.window-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/reportonly-default-src.https.window.js.sub.headers: Avoid domains.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/reportonly-script-src-elem.https.window-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/reportonly-script-src-elem.https.window.js.sub.headers: Avoid domains.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/reportonly-script-src-none.https.window-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/reportonly-script-src-none.https.window.js.sub.headers: Avoid domains.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/reportonly-script-src.https.window-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/reportonly-script-src.https.window.js.sub.headers: Avoid domains.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/resources/report-hash-test-runner.sub.js:
(report_hash_test): Avoid domains and add error handling.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/script-src-elem.https.window-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/script-src-elem.https.window.js.sub.headers: Avoid domains.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/script-src-sha512.https.window-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/script-src-sha512.https.window.js.sub.headers: Avoid domains.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/script-src.https.window-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/report-hash/script-src.https.window.js.sub.headers: Avoid domains.
* LayoutTests/platform/mac-wk1/TestExpectations: Expect failures on wk1 - https://bugs.webkit.org/show_bug.cgi?id=285098.
* Source/WebCore/Modules/reporting/ViolationReportType.h: Add
  CSPHashReport type.
* Source/WebCore/bindings/js/ScriptModuleLoader.cpp:
(WebCore::ScriptModuleLoader::notifyFinished): rename
matchIntegrityMetadata call.
* Source/WebCore/dom/LoadableClassicScript.cpp:
(WebCore::LoadableNonModuleScriptBase::notifyFinished): rename
matchIntegrityMetadata call.
* Source/WebCore/html/HTMLLinkElement.cpp:
(WebCore::HTMLLinkElement::setCSSStyleSheet): rename
matchIntegrityMetadata call.
* Source/WebCore/loader/DocumentThreadableLoader.cpp:
(WebCore::DocumentThreadableLoader::didFinishLoading): rename
matchIntegrityMetadata call.
* Source/WebCore/loader/PingLoader.cpp:
(WebCore::PingLoader::sendViolationReport): Handle CSPHashReport.
* Source/WebCore/loader/SubresourceIntegrity.cpp:
(WebCore::createReportFormData): Create a report.
(WebCore::addHashPrefix): Add a prefix to the reported value, based on
the algorithm enum value.
(WebCore::findStrongestAlgorithm): Get the strongest algorithm is
a HashAlgorithmSet.
(WebCore::reportHashesIfNeeded): Potentially report hashes for a resource.
(WebCore::matchIntegrityMetadataSlow): implements the
matchIntegrityMetadata logic.
* Source/WebCore/loader/SubresourceIntegrity.h:
(WebCore::matchIntegrityMetadata): Adjust condition and call hash
reporting.
* Source/WebCore/loader/cache/CachedResource.h:
(WebCore::CachedResource::loader const): Make it a const.
(WebCore::CachedResource::setIsHashReportingNeeded): Setter for
isHashReportingNeeded.
(WebCore::CachedResource::isHashReportingNeeded const): Getter for
isHashReportingNeeded.
(WebCore::CachedResource::loader): Deleted.
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::requestResource): Set
isHashReportingNeeded on the resource.
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
(WebCore::ContentSecurityPolicy::hashesToReport): Get the hashes to
report from the different CSP policies.
* Source/WebCore/page/csp/ContentSecurityPolicy.h: Define
  HashAlgorithmSet and HashAlgorithmSetCollection.
* Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.cpp:
(WebCore::ContentSecurityPolicyDirectiveList::hashReportDirectiveForScript const): Get the directives for script hash reporting.
(WebCore::ContentSecurityPolicyDirectiveList::reportHash const): Return
the HashAlgorithmSet for the script reporting directive.
* Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.h: Add
  reportHash and hashReportDirectiveForScript.
* Source/WebCore/page/csp/ContentSecurityPolicySourceList.cpp:
(WebCore::ContentSecurityPolicySourceList::parseSource): Parse the hash
reporting keywords and set the appropriate algorithms in the
HashAlgorithmSet.
* Source/WebCore/page/csp/ContentSecurityPolicySourceList.h: Add
  a HashAlgorithmSet.
(WebCore::ContentSecurityPolicySourceList::reportHash const): Getter for
the HashAlgorithmSet.
* Source/WebCore/page/csp/ContentSecurityPolicySourceListDirective.h:
(WebCore::ContentSecurityPolicySourceListDirective::reportHash const):
Pipe the HashAlgorithmSet from the sourceList.

Canonical link: https://commits.webkit.org/288506@main
